Appveyor building `hab` and testing ported crates

smurawski opened the pull request:
I moved the execution of the build and test cycles into a powershell script (`./support/ci/appveyor.ps1`) which is called by (`./support/ci/appveyor.bat`).  

We use a batch file to start things off because AppVeyor's PowerShell host treats stderr from native executables (like `cargo`) as errors on the error stream and adds all sorts of ugly output.

Calling PowerShell.exe from a batch file, while an odd hack, works to keep the output looking good but have the better scripting experience of PowerShell.

Now, the build will check if any files have changed in the ported crates and, if so, it'll run cargo test against all ported crates and then build `hab.exe` and run `hab --version` to validate `hab.exe` can run successfully.

Builds can be forced by setting a value to $env:HAB_FORCE_BUILD and tests can be forced with $env:HAB_FORCE_TEST

Signed-off-by: Steven Murawski <steven.murawski@gmail.com>

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

mwrock commented on appveyor.yml:
```diff
@@ -23,5 +30,4 @@ install:
 test_script:
-  - cd c:\projects\habitat\components\hab
-  - cargo build --verbose
+  - appveyor.bat
```
Would be great to see this in powershell because `%%A` makes me reactively stab my eyes :)

smurawski replied:
yeah, the problem is the output from cargo via powershell makes me want to stab my eyes.

mwrock replied:
haha. ok fair enough!

mwrock commented on appveyor.yml:
```diff
@@ -15,6 +16,13 @@ skip_tags: true
+      hab_components: "core;http-client;builder-protocol;builder-depot-client;common"
+
 install:
   - ps: ./components/hab/win/win-build.ps1
```
Not a bad idea to add a `& /path/to/hab.exe --version`

smurawski replied:
After the build?  It'll only exist after the cargo build, not in the test cycle.

smurawski commented on appveyor.yml:
```diff
@@ -15,6 +16,13 @@ skip_tags: true
+environment:
+  matrix:
+    - hab_build_action: build
+      hab_components: hab
```
Correct (though I should add hab to the collection of for the test run).

reset replied:
Yeah I'd add `hab` to the list of components to be tested. Do you think there's much value in a matrix just ensuring the component builds? The cargo test task will build the software in addition to running the test suite. It's true that we could be using `cfg(test)` for conditional compilation outside of just marking modules as test modules, but we don't really do that.

Whatever you feel is most appropriate just make sure we run the test suite on all components ;)

smurawski commented:
@mwrock @reset I changed the direction a bit. Matt, your eyes will bleed less as I've moved to a PowerShell script for the conditional logic.  Jamie, we are building hab.exe and running tests against all the crates.  

There is no benefit at this time (and, in fact, an impediment thanks to cargo workspaces) in splitting the builds onto multiple machines.

mwrock commented on appveyor.yml:
```diff
@@ -15,6 +16,11 @@ skip_tags: true
 
+environment:
+  matrix:
+    - hab_build_action: "test;build"
```
Any reason why not to break out the matrix like:

```
- hab_build_action: "build"
- hab_components: "hab"

- hab_build_action: "test"
- hab_components: "hab"

- hab_build_action: "test"
- hab_components: "http-client"

etc, etc
```

Something like this would allow the appveyor.ps1 to be much simpler (no looping or string parsing) and the test actions to be more clear.

smurawski replied:
yeah, it's WAAAY slower.  
* sharing a workspace for each build step allows single pass compilation of dependencies
* if we don't want the build matrix to progress serially, there's a dollar cost ($100/mo just to get two concurrent jobs).

smurawski commented:
@thesentinels approve

thesentinels commented:
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

thesentinels commented:
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/168670790).

thesentinels commented:
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/168670790).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

![gif-keyboard-3280869874741411265](https://cloud.githubusercontent.com/assets/4304/17755674/1e577b82-6490-11e6-96b7-1663d283c824.gif)
